core/src/replay_stage.rs

```diff
@@ -199,11 +203,43 @@ impl ReplayStage {
             snapshot_package_sender,
             block_commitment_cache,
             transaction_status_sender,
+            tower_snapshot_path,
+            allow_missing_tower_state,
         } = config;
 
         let (root_bank_sender, root_bank_receiver) = channel();
         trace!("replay stage");
-        let mut tower = Tower::new(&my_pubkey, &vote_account, &bank_forks.read().unwrap());
+        let tower = Tower::reload_from_file(&tower_snapshot_path, &my_pubkey, &vote_account);
+
+        let mut tower = match tower {
+            Ok(tower) => tower,
+            Err(e) => {
+                let bank_forks_lock = bank_forks.read().unwrap();
+                if let Some(account) = bank_forks_lock
+                    .get(bank_forks_lock.root())
+                    .expect("Failed to get root bank")
+                    .get_account(&vote_account)
+                {
+                    if let Some(vote_state) = VoteState::from(&account) {
+                        if !vote_state.votes.is_empty() {
+                            if allow_missing_tower_state {
+                                error!("Found initialized vote account with votes, but opening saved tower failed with {:?}", e);
+                            } else {
+                                panic!("Found initialized vote account with votes, but opening saved tower failed with {:?}", e);
+                            }
+                        }
+                    }
+                }
+                Tower::new(&my_pubkey, &vote_account, &bank_forks.read().unwrap())
+            }
+        };
+
+        // Tower gets written to before blocktree, so the root may be missing from blocktree
+        if let Some(tower_root) = tower.root() {
+            if !blocktree.is_root(tower_root) {
+                blocktree.set_roots(&[tower_root]).unwrap();
```

Comment:
This won't set the roots for roots between the latest blocktree root and the tower root. 

For instance imagine the latest blocktree root is 3, and your `BankForks` looks like: 

`3, 4, 5, 6, 7`

The validator roots `7`, and saves the tower. At this point, 4, 5, 6, 7 all need to be save to blocktree's `roots` column family. However if there is a crash before `set_root(&[4, 5, 6, 7])` can be called, then these need to all be restored on startup. Here the logic will only call `set_root(&[7]`.

I think because blocktree_processor will replay all the existing ledger slots when rebuilding BankForks on restart, that chain of slots `4, 5, 6, 7` will exist in BankForks. Thus, iff we call `bank_forks.set_root(tower_root)` here, then BankForks should commit that chain to the rooted slots in blocktree.

A test showing this behavior would be nice. :D

Reply:
@t-nelson Along the same vein, I think we need to reason through the below cases. In the below cases, assume that validator is rebooting, and that they crashed directly after storing the root in tower, but before they persisted the root to blocktree. (Aka root in tower > last root in blocktree). After we've synchronized blocktree and tower using the procedure listed in the comment above:

1) What happens if the validator boots up from a snapshot with root < tower root?

Thoughts: In this case, assume snapshot verification has verified that snapshot root is an ancestor of the tower root. Then the only remaining problem is that on boot from blocktree processor, because the root in blocktree < tower root at the time, the boot procedure did not call set_root() on the latest tower root (logic on boot to set root is here: https://github.com/solana-labs/solana/blob/master/ledger/src/blocktree_processor.rs#L547).

Now ReplayStage can potentially play forks with ancestors earlier than the root, and this breaks assumptions in the code that BankForks doesn't track ancestors with roots earlier than locktower root.

Solution: Move the logic to synchronize blocktree root and tower root to blocktree_processor on boot, so that the boot logic has the latest root information. 

TODOS: Add test case for this.

2) What happens if the validator boots up from a snapshot with root `S` > tower root `T` here

Thoughts: This means we will boot from a root farther than the validator's last known root. blocktree_processor handles this on boot by only replaying forks that are descendants of `S`. However, this means that there will be entries in the validator's `Tower` that don't exist as banks in `BankForks`. We should double check that there are no assumptions in current `consensus.rs` that assume an entry in tower must exist in `BankForks` (for instance unwrapping a lookup for a bank in `BankForks`)

Solution:  I think we should still synchronize blocktree root and tower root as suggested in 1)

TODOS: Add test case for this, check `consensus.rs` for assumptions that an entry in tower must exist in `BankForks`

Reply:
Thanks for the thoughts!  I've been thinking the same, that we should reconcile the local state _before_ worrying about the snapshot and that it should be an independent step.  Seems like a PR all its own, wdyt?

Reply:
yup sounds good to me! Can you reference the design here in that PR so we can keep track of the thought process? Thanks!

Reply:
So I've got the `Blockstore` / `Tower` root sync in https://github.com/solana-labs/solana/pull/7436/commits/52da84ac3b169b8691712f624c15e528e7f697dc

Regarding the `Solution` from 1), https://github.com/solana-labs/solana/blob/73a278dc647d12cf43a6c14dccfd1edb42cc0ee5/core/src/validator.rs#L176 seems like the earliest place to do the sync.  We'll have to restore the `Tower` clear up here and pass it down, but I don't see any problems with that.

Reply:
Actually, might want to put the `Tower` together before the `new_banks_from_blockstore()` call, pass the `Tower` in and sync immediately after opening the `Blockstore`

Reply:
@carllin https://github.com/solana-labs/solana/pull/7436/commits/6f46267f3bafe10405c2d9ca3251529a2f49770d should take care of getting the `Blockstore` roots in sync with the `Tower` ASAP.  I just need to get the tests written